docs/plugins/testing_workshop_docs/Article-3-readers-and-fixtures.md

```diff
@@ -0,0 +1,152 @@
+
+```python
+    # tmp_path is a pytest fixture  
+    def test_get_reader_returns_callable(tmp_path):  
+        """Calling get_reader on numpy file returns callable"""  
+   
+        # write some fake data  
+        my_test_file = str(tmp_path / "myfile.npy")  
+        original_data = np.random.rand(20, 20)  
+        np.save(my_test_file, original_data)  
+ 
+        # try to read it back in  
+        reader = napari_get_reader(my_test_file)  
+        assert callable(reader)
+```
+ 
+Running the command `pytest .` in the root directory of the plugin, we discover all the tests it recognizes as tests. It should recognize `test_reader.py` because it's a test file, prefixed with the word test. `test_reader.py` was found and passed the test. 
+
+![pytest passed](../../images/napari_plugins_3rd_pytest_passed.png)
+  
+If the file did not end in `.npy` the test would fail because what was returned wasn't callable. This code has been modified to produce an error:  
+```python    
+    # tmp_path is a pytest fixture  
+    def test_get_reader_returns_callable(tmp_path):  
+        """Calling get_reader on numpy file returns callable"""  
+ 
+        # write some fake data
+        my_test_file = str(tmp_path / "myfile.np") # note ends in .np  
+        original_data = np.random.rand(20, 20)  
+        np.save(my_test_file, original_data)  
+ 
+        # try to read it back in  
+        reader = napari_get_reader(my_test_file)  
+        assert callable(reader)  
+```
+Once we run `pytest` we can see that it traced back that the callable of `reader` is `False` and it has filled in the fact that `reader` at the time of the assertion was `None`. This is useful in debugging. 
+
+![test_get_reader_returns_callable Failed](../../images/napari_plugins_4th_test_get_reader_returns_callable-failed.png)
+
+## Custom fixtures and round-trip tests
+Next, we test to see if this function reads the data. This is a round-trip test. We will create a fixture to write the data to make things easier for ourselves. This fixture will be called [test_reader_round_trip](https://github.com/DragaDoncila/plugin-tests/blob/effb32d6e3b191ad83e69813b26ae8695210f5ad/src/plugin_tests/_tests/test_reader.py#L39).   
+  
+Whatever is returned out of a `@pytest.fixture` decorated function is passed as an argument with the name of the fixture, to the test. We are going to call this `pytest.fixture` decorated function `write_im_to_file`, as defined on line 6. We’re going to give this fixture the `tmp_path` fixture. Fixtures can use fixtures. 
+
+`write_im_to_file` returns a function we will call `write_func` that we can pass a path to and have it write the numpy file to that path. `write_func` is defined inside `write_im_to_file` because it’s not needed anywhere else. When we use `write_im_to_file` inside this test it will be referred to as `write_func`. We can call it with a file name, save the data, and read the data back.   
```

Comment:
```suggestion
`write_im_to_file` returns a function we will call `write_func` that we can pass a path to and have it write the numpy file to that path. `write_func` is defined inside `write_im_to_file` because it’s not needed anywhere else. When we use `write_im_to_file` inside this test it will run `write_func`.
```